Storage/SQL/SQLiteBookmarksModel.swift

```diff
@@ -280,7 +275,27 @@ extension SQLiteBookmarks {
             "ORDER BY title ASC",
             ].joinWithSeparator(" ")
 
-        return self.db.runQuery(withIcon, args: args, factory: BookmarkFactory.factory)
+        let sql = includeIcon ? withIcon : values + " ORDER BY title ASC"
+        return isEditableForDirection(direction) >>== { editable in
+            args.insert((editable ? 1 : 0) as AnyObject, atIndex: 0)
+            return self.db.runQuery(sql, args: args, factory: BookmarkFactory.factory)
+        }
+    }
+
+    private func isBufferEmpty() -> Deferred<Maybe<Bool>> {
+        return self.db.queryReturnsNoResults("SELECT 1 FROM \(TableBookmarksBuffer)")
+    }
+
+    private func isMirrorEmpty() -> Deferred<Maybe<Bool>> {
+        return self.db.queryReturnsNoResults("SELECT 1 FROM \(TableBookmarksMirror)")
+    }
+
+    func isEditableForDirection(direction: Direction) -> Deferred<Maybe<Bool>> {
+        return accumulate([isBufferEmpty, isMirrorEmpty]) >>== { checks in
+            let bufferEmpty = checks[0]
+            let mirrorEmpty = checks[1]
+            return deferMaybe(bufferEmpty && direction == .Local ? true : mirrorEmpty)
```

Comment:
If `direction != .Local`, you don't even need to run the first query, right?

But even so, here you only need one query!

```
SELECT
ifnull((SELECT 1 FROM \(TableBookmarksBuffer)), 0) AS hasBuffer,
ifnull((SELECT 1 FROM \(TableBookmarksMirror)), 0) AS hasMirror
```

gives you two columns, one for each table.